### PR TITLE
Use EncodableToTLV in im_responder

### PR DESCRIPTION
**Pull request by andy31415**
This makes the API cleaner and safer (have snapshot & rollback capabilities, less use of prepare/finish).

Goal is to start treating PrepareCommand/GetCommandDataIBTLVWriter/FinishCommand as private APIs.

### CODE REVIEW

**Review by yunhanw-google — APPROVED** [review]

File: src/app/tests/integration/chip_im_responder.cpp [hunk 1]
```diff
@@ -50,6 +50,25 @@
 namespace chip {
 namespace app {
 
+namespace {
+
+class TestTLVDataEncoder : public chip::app::DataModel::EncodableToTLV
```
Comment (bzbarsky-apple) [inline src/app/tests/integration/chip_im_responder.cpp:55]:
```suggestion
class TestTLVDataEncoder : public DataModel::EncodableToTLV
```
This is all inside chip::app.

File: src/app/tests/integration/chip_im_responder.cpp [hunk 2]
```diff
@@ -104,17 +123,8 @@ void DispatchSingleClusterCommand(const ConcreteCommandPath & aRequestCommandPat
...
-        ReturnOnFailure(apCommandObj->PrepareInvokeResponseCommand(path, prepareParams));
-
-        writer = apCommandObj->GetCommandDataIBTLVWriter();
-        ReturnOnFailure(writer->Put(chip::TLV::ContextTag(kTestFieldId1), kTestFieldValue1));
-
-        ReturnOnFailure(writer->Put(chip::TLV::ContextTag(kTestFieldId2), kTestFieldValue2));
-
-        ReturnOnFailure(apCommandObj->FinishCommand());
+        TestTLVDataEncoder testData;
+        apCommandObj->AddResponse(path, kTestCommandId, testData);
```
Comment (bzbarsky-apple) [inline src/app/tests/integration/chip_im_responder.cpp:127]: So here's a question....  AddResponse will always need to take a thing that encodes a struct start/end, no?  Seems like we should consider a `StructEncodableToTLV` thing which has a single impl of those and then calls some other virtual method it calls that can be overridden to encode the struct fields...  Not part of this PR, but both this and the network commissioning bits could use it.

**Review by bzbarsky-apple — APPROVED** [review]
